File: javascript/simple-linked-list/simple-linked-list.js
export class Element {
    constructor(value) {
        this.value = value;
        this.next = null;
    }
}

export class List {
    constructor(init) {
        this.head = null;
        this.length = 0;
        if (init) {
            for (let value of init) {
                this.add(new Element(value));
            }
        }
    }

    add(element) {
        element.next = this.head;
        this.head = element;
        this.length++;
    }

    toArray() {
        const arr = [];
        let node = this.head;
        while (node !== null) {
            arr.push(node.value);
            node = node.next;
        }
        return arr;
    }

    reverse() {
        return new List(this.toArray());
    }
}

File: ruby/simple-linked-list/_iterations/1/simple_linked_list.rb
class Element < Struct.new(:datum, :next); end

class SimpleLinkedList
  def initialize(enum = [])
    enum.each { push Element.new(_1) }
  end

  def push(element)
    element.next = @head
    @head = element
    self
  end

  def pop
    element = @head
    @head = @head.next unless @head.nil?
    element
  end

  def each_element
    return to_enum(:each_element) unless block_given?
    
    element = @head
    until element.nil?
      yield element
      element = element.next
    end
  end

  def to_a
    each_element.map(&:datum)
  end

  def reverse!
    elements = each_element.to_a.reverse
    elements.zip(elements[1..-1]).each { |a, b| a.next = b }
    @head = elements.first
    self
  end
end
